### PR TITLE
feat(inputs.smartctl): Introduce smartctl JSON  input plugin

### PR DESCRIPTION
**Pull request by powersj**
## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #

### CODE REVIEW

**Comment by nlgranger** [comment]
Hi! To answer your question in the issue, here is the command which returns exit code 4
`sudo smartctl --json --all /dev/bus/6 --device megaraid,14 --nocheck=standby`
[smartctl_megaraid.txt](https://github.com/influxdata/telegraf/files/14791795/smartctl_megaraid.txt)

**Comment by nlgranger** [comment]
I think the plugin only collects data from the last disk of the raid controller.

```
$ sudo smartctl --scan
/dev/sda -d scsi # /dev/sda, SCSI device
/dev/sdb -d scsi # /dev/sdb, SCSI device
/dev/sdc -d sat # /dev/sdc [SAT], ATA device
/dev/bus/6 -d megaraid,8 # /dev/bus/6 [megaraid_disk_08], SCSI device
/dev/bus/6 -d megaraid,9 # /dev/bus/6 [megaraid_disk_09], SCSI device
/dev/bus/6 -d megaraid,10 # /dev/bus/6 [megaraid_disk_10], SCSI device
/dev/bus/6 -d megaraid,11 # /dev/bus/6 [megaraid_disk_11], SCSI device
/dev/bus/6 -d megaraid,12 # /dev/bus/6 [megaraid_disk_12], SCSI device
/dev/bus/6 -d megaraid,13 # /dev/bus/6 [megaraid_disk_13], SCSI device
/dev/bus/6 -d megaraid,14 # /dev/bus/6 [megaraid_disk_14], SCSI device
```
</details>

Only disk 14 is mentioned in the telegraf logs:

```
smartctl,host=**********,model=LSI\ MR9361-8i,name=/dev/sda,serial=00dfc2930dda29b821305dcc0cb00506,type=scsi firmware="",capacity=255516999680i,health_ok=false,temperature=0i 1711745550000000000
smartctl,host=**********,model=ST6000NM0115-1YZ110,name=/dev/bus/6,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d capacity=6001175126016i,health_ok=true,temperature=25i,firmware="SN04" 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Raw_Read_Error_Rate,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d value=83i,raw_value=185123688i,worst=64i,threshold=44i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Spin_Up_Time,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=91i,threshold=0i,value=91i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Start_Stop_Count,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=62i,worst=100i,threshold=20i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Reallocated_Sector_Ct,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=100i,threshold=10i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Seek_Error_Rate,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=2836411902i,worst=60i,threshold=45i,value=95i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Power_On_Hours,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=44345i,worst=50i,threshold=0i,value=50i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Spin_Retry_Count,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d worst=100i,threshold=97i,value=100i,raw_value=0i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Power_Cycle_Count,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=62i,worst=100i,threshold=20i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=End-to-End_Error,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=100i,threshold=99i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Reported_Uncorrect,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d worst=100i,threshold=0i,value=100i,raw_value=0i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Command_Timeout,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=100i,threshold=0i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=High_Fly_Writes,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d worst=100i,threshold=0i,value=100i,raw_value=0i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Airflow_Temperature_Cel,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=504627225i,worst=64i,threshold=40i,value=75i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=G-Sense_Error_Rate,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=3296i,worst=99i,threshold=0i,value=99i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Power-Off_Retract_Count,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=1866i,worst=100i,threshold=0i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Load_Cycle_Count,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d worst=100i,threshold=0i,value=100i,raw_value=1895i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Temperature_Celsius,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d value=25i,raw_value=30064771097i,worst=40i,threshold=0i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Hardware_ECC_Recovered,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d value=83i,raw_value=185123688i,worst=64i,threshold=0i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Current_Pending_Sector,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=100i,threshold=0i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Offline_Uncorrectable,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=100i,threshold=0i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=UDMA_CRC_Error_Count,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=0i,worst=200i,threshold=0i,value=200i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Head_Flying_Hours,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=8331506409844008i,worst=253i,threshold=0i,value=100i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Total_LBAs_Written,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d worst=253i,threshold=0i,value=100i,raw_value=28115177774i 1711745550000000000
smartctl_attributes,host=**********,model=ST6000NM0115-1YZ110,name=Total_LBAs_Read,serial=ZAD2C11G,type=sat+megaraid\,14,wwn=5000c500a496983d raw_value=3197497259616i,worst=253i,threshold=0i,value=100i 1711745550000000000
smartctl,host=**********,model=LSI\ MR9361-8i,name=/dev/sdb,serial=00aa72f30f022ab821305dcc0cb00506,type=scsi capacity=18001818550272i,health_ok=false,temperature=0i,firmware="" 1711745550000000000
smartctl,host=**********,name=/dev/sdc,type=sat temperature=0i,firmware="",capacity=0i,health_ok=false 1711745550000000000
```

File: plugins/inputs/smartctl/smartctl_scan.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+	var scan smartctlScanJSON
+	if err := json.Unmarshal(out, &scan); err != nil {
+		return nil, fmt.Errorf("error unmarshalling smartctl scan output: %w", err)
+	}
+
+	devices := make(map[string]string, len(scan.Devices))
+	for _, device := range scan.Devices {
+		if s.deviceFilter.Match(device.Name) {
+			devices[device.Name] = device.Type
```
[the remark this answers is not in the document]
Comment (powersj) [reply plugins/inputs/smartctl/smartctl_scan.go:30]: yep! I assumed names would be unique just like we already incorrectly do today :D I'll update on Monday.

**Comment by nlgranger** [comment]
Adding scan result with megaraid controller in case you want to make a test out of it.
[smartctl_scan_megaraid.txt](https://github.com/influxdata/telegraf/files/14808634/smartctl_scan_megaraid.txt)

**Comment by powersj** [comment]
> Adding scan result with megaraid controller in case you want to make a test out of it. [smartctl_scan_megaraid.txt](https://github.com/influxdata/telegraf/files/14808634/smartctl_scan_megaraid.txt)

Yes, thank you very much!

**Comment by powersj** [comment]
@nlgranger,

I've pushed an update that does not assume unique names during scan. The next set of artifacts after this message will contain a build with those changes. If you could once again give it a shot and let me know if all the data is collected, I would really appreciate it!

Thanks again!

**Comment by nlgranger** [comment]
It work perfectly now!

> @nlgranger,
> 
> I've pushed an update that does not assume unique names during scan. The next set of artifacts after this message will contain a build with those changes. If you could once again give it a shot and let me know if all the data is collected, I would really appreciate it!
> 
> Thanks again!

**Comment by powersj** [comment]
Thanks for confirming!

**Comment by srebhan** [comment]
@powersj FYI: there is a pure-go library for accessing SMART information: https://github.com/anatol/smart.go

**Comment by DEvil0000** [comment]
I am comming from #15095. I gave the build above from the bot a spin. result however is it is working for normal disks but not showing any smart data for the raid disks. Same config and machine as in #15095 - so attribute read is not enabled in config.

**Comment by powersj** [comment]
> however is it is working for normal disks but not showing any smart data for the raid disks. Same config and machine as in https://github.com/influxdata/telegraf/issues/15095

Hopefully you are not using the same config, as this PR introduces a new plugin called `inputs.smartctl` to parse the JSON output.

If you still have issues after updating the input name, please provide the results of:

* `smartctl --json --scan`

And the device scan from the device you are not seeing:

* `smartctl --json --all $DEVICE --device $TYPE`

**Comment by DEvil0000** [comment]
thanks for the config change hint. I just missed that. while i just added the output above
works fine for me then.

the path for me however is `/usr/sbin/smartctl` and not `/usr/bin/smartctl` - you may want consider a change there.
can I get a enable/disable option for the attributes - I am just interested in health normally (`-H` parameter)?

**Comment by powersj** [comment]
> works fine for me then.

Thanks for confirming!

> the path for me however is /usr/sbin/smartctl and not /usr/bin/smartctl - you may want consider a change there.

There is a config option for this ;)

```toml
    ## Optionally specify the path to the smartctl executable
    # path = "/usr/bin/smartctl"
```

> can I get a enable/disable option for the attributes 

You can use [metric filtering](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#metric-filtering) to remove unwanted metrics. For example, using the `namedrop` config parameter prevents metrics with a certain name leaving the metric: `namedrop = ["smartctl_attributes"]

**Comment by DEvil0000** [comment]
btw: thanks for the work you put in this!

another thing I realized is that the smart plugin in newer versions used to call `nvme` cli tool as well for some details. Not sure if that actually matters or you want to include this as well.

> There is a config option for this ;)

That is what I used. Pointing to `bin` however just looks like a odd default to me since I do not know any distro packaging it to this path.

> can I get a enable/disable option for the attributes

sure I can filter but this option would have been an more easy config and may reduce computational load.

**Comment by nlgranger** [comment]
> I do not know any distro packaging it to this path.

Archlinux! (BTW™)

**Comment by powersj** [comment]
> Not sure if that actually matters or you want to include this as well.

Yeah this is part of the reason why this is a new plugin. I specifically did not want to continue doing this. The nvme calls should have been a second plugin.

>> I do not know any distro packaging it to this path.
> Archlinux! (BTW™)

hahah exactly :) I copied that path over from the previous plugin, but if the bulk of distros use `/usr/sbin` then I can update that to the default. I know debian/ubuntu use `/usr/sbin`. If Fedora does the same, then I think that would be ok to switch.

**Comment by nlgranger** [comment]
> If Fedora does the same, then I think that would be ok to switch.

RHEL does.

**Comment by powersj** [comment]
@srebhan,

> @powersj FYI: there is a pure-go library for accessing SMART information: https://github.com/anatol/smart.go

Thanks for finding this. I took a look at this and have the following comments:

1. The library does not scan or find devices for us. The examples use [github.com/jaypipes/ghw](https://github.com/jaypipes/ghw) to discover disks. I would rather us continue to use the scan JSON output from the command itself as it is a bit easier to ask users for that output if issues are found.
2. smartmontools does exists for other platforms while that library is linux on. They do have some darwin files, but require cgo. While I have not tried this, if the JSON output works on those platforms they this plugin should work on them as well.
3. Like the nvidia_smi plugin I prefer to parse the actual data from the tool itself and allow us to easily add new metrics or fields that may get requested. I do not want to depend on another library (i.e. gopsutil) in another plugin.
4. The commands to get smart data depend on the type, and may not cover all types. For example the megaraid in the above user's comments may not actually be covered. We may be able to switch this to SCSI, but it would be logic we have to maintain.

**Review by srebhan — APPROVED** [review]
Thanks @powersj! The code looks good, my only question is how this relates to `inputs.smart`? Is this a replacement? If not, what are the differences?

**Comment by powersj** [comment]
> how this relates to inputs.smart? Is this a replacement? If not, what are the differences?

Should I add the following to the readme?

Smartctl is focused only on collecting data from smartctl via the JSON output it provides via the scan and collection commands.

The smart plugin parses the raw output of the smartctl using a variety of regex. It can also optionally collect additional details via the nvme command including some vendor specific details.

I don't fully consider smartctl as a replacement, given it lacks the additional intel nvme collection/details, but for users who report issues using smart in the future, I would absolutely point them at smartctl first.

**Comment by srebhan** [comment]
@powersj after sleeping over it, it would be nice to add this statement to the readme. I guess this will also clarify things for the user.

**Comment by powersj** [comment]
> @powersj after sleeping over it, it would be nice to add this statement to the readme. I guess this will also clarify things for the user.

done, give it a read and let me know if I should add something to the smart plugin as well.

**Review by srebhan — APPROVED** [review]
Awesome! Thanks @powersj!

**Review by DStrand1 — APPROVED** [review]

**Pull request merged**